commons-core/src/test/scala/com/avsystem/commons/serialization/json/JsonStringInputOutputTest.scala

```diff
@@ -384,4 +385,29 @@ class JsonStringInputOutputTest extends AnyFunSuite with SerializationTestUtils
   test("read escaped slash") {
     assert(JsonStringInput.read[String](""""a\/b"""") == "a/b")
   }
+
+  test("peeking fields - empty object") {
+    val json = """{}"""
+    val oi = new JsonStringInput(new JsonReader(json)).readObject()
+    assert(oi.peekField("x").isEmpty)
+    assert(!oi.hasNext)
+  }
+
+  test("peeking fields - non empty object") {
+    val json = """{"a": 25, "b": true, "c": [1,2,3], "d": {}}"""
+    val oi = new JsonStringInput(new JsonReader(json)).readObject()
+    assert(oi.peekField("a").map(GenCodec.read[Int]).contains(25))
+    assert(oi.peekField("b").map(GenCodec.read[Boolean]).contains(true))
+    assert(oi.nextField().fieldName == "a") // peeking should not affect `nextField`
+    assert(oi.peekField("d").map(GenCodec.read[Map[String, String]]).contains(Map.empty))
+    assert(oi.peekField("c").map(GenCodec.read[List[Int]]).contains(List(1, 2, 3)))
+    assert(oi.peekField("x").isEmpty)
+    assert(oi.peekField("a").map(GenCodec.read[Int]).contains(25))
+    assert(oi.nextField().fieldName == "b") // peeking should not affect `nextField`
+  }
+
+  test("reading flat sealed hierarchy with changed field order") {
```

Comment:
Yeah, this is the result of incrementally adding stuff
* at first, there was no such thing as `peekField`, `Input`s & `Output`s were strictly required to keep the order
* `@outOfOrder` was added as a band-aid specifically for the fact that Mongo always returns `_id` as first field (otherwise keeping the field order properly)
* `peekField` was added then but it's not supported by all inputs/outputs so you can only rely on it when using specific input/output implementation

Overall, the initial design assumed that `Input` will _only_ read data that was written by its corresponding `Output`. So _technically_, `JsonStringInput` should not be used to read user data, e.g. in REST. Reality is, of course, different.

Reply:
Yeah, I guess it would make sense to add a general recommendation in gencodec.md about using `outOfOrder` vs peeking to avoid confusion.

Reply:
Damn, I want to rewrite this entire doc. It's old.